qiskit/circuit/library/arithmetic/adders/qft_adder.py

```diff
@@ -0,0 +1,109 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Compute the sum of two qubit registers using QFT."""
+
+import numpy as np
+
+from qiskit.circuit.quantumregister import QuantumRegister
+from qiskit.circuit.library.basis_change import QFT
+
+from .adder import Adder
+
+
+class QFTAdder(Adder):
+    r"""A circuit that uses QFT to perform in-place addition on two qubit registers.
+
+    For registers with :math:`n` qubits, the QFT adder can perform addition modulo
+    :math:`2^n` (with `modular=True`) or ordinary addition by adding a carry qubits (with
+    `modular=False`).
+
+    As an example, a non-modular QFT adder circuit that performs addition on two 2-qubit sized
+    registers is as follows:
+
+    .. parsed-literal::
+
+         a_0:   ─────────■──────■────────────────────────■────────────────
+                         │      │                        │
+         a_1:   ─────────┼──────┼────────■──────■────────┼────────────────
+                ┌──────┐ │P(π)  │        │      │        │       ┌───────┐
+         b_0:   ┤0     ├─■──────┼────────┼──────┼────────┼───────┤0      ├
+                │      │        │P(π/2)  │P(π)  │        │       │       │
+         b_1:   ┤1 qft ├────────■────────■──────┼────────┼───────┤1 iqft ├
+                │      │                        │P(π/2)  │P(π/4) │       │
+        cout_0: ┤2     ├────────────────────────■────────■───────┤2      ├
+                └──────┘                                         └───────┘
+
```

Comment:
Why is there a `cout` bit here? The paper's figure doesn't have it. My [old implementation](https://github.com/ajavadia/qiskit-terra/blob/Demo/demo/Circuit%20Library%20Demo.ipynb) of this also didn't have it and it seemed to work.

```


                                                                                        
b_3: ─────────■─────────────────────────────────────────────────────────────────────────
              │                                                                         
b_2: ─────────┼────■─────────────────────■──────────────────────────────────────────────
              │    │                     │                                              
b_1: ─────────┼────┼──────■──────────────┼────■──────────────■──────────────────────────
              │    │      │              │    │              │                          
b_0: ─────────┼────┼──────┼──────■───────┼────┼──────■───────┼────■───────■─────────────
     ┌──────┐ │pi  │pi/2  │pi/4  │pi/8   │    │      │       │    │       │    ┌───────┐
a_3: ┤3     ├─■────■──────■──────■───────┼────┼──────┼───────┼────┼───────┼────┤3      ├
     │      │                            │pi  │pi/2  │pi/4   │    │       │    │       │
a_2: ┤2     ├────────────────────────────■────■──────■───────┼────┼───────┼────┤2      ├
     │  qft │                                                │pi  │pi/2   │    │  iqft │
a_1: ┤1     ├────────────────────────────────────────────────■────■───────┼────┤1      ├
     │      │                                                             │pi  │       │
a_0: ┤0     ├─────────────────────────────────────────────────────────────■────┤0      ├
     └──────┘                                                                  └───────┘
```

Reply:
The circuit you show above is a modular addition, but you can generalize the QFT adder to a "proper" non-modular addition by adding a carry qubit and using this implementation: Chapter 4 in https://arxiv.org/pdf/1411.5949.pdf

Reply:
ok. the pattern of 2-qubit gates looks a bit strange, I can't understand what the pattern is. But I guess it's working.